### PR TITLE
add rule E3017 to check when a prop is required based on a value

### PR DESCRIPTION
**Pull request by kddejong**
*Issue #, if available:*
#1744 
*Description of changes:*
- Add rule E3017 to validate when a property is required based on another property having a certain value

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/cfnlint/data/AdditionalSpecs/RequiredBasedOnValue.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+                        "AWS::MSK::Cluster"
+                    ],
+                    "RequiredProperties": [
+                        "StartingPosition"
+                    ]
+                }
+            ]
+        }
+    }
+}
```
Comment (PatMyron) [inline src/cfnlint/data/AdditionalSpecs/RequiredBasedOnValue.json:22]: Like reading directly from [`AdditionalSpecs`](https://github.com/aws-cloudformation/cfn-python-lint/blob/master/src/cfnlint/data/AdditionalSpecs) and avoiding the patching process of [`ExtendedSpecs`](https://github.com/aws-cloudformation/cfn-python-lint/blame/master/src/cfnlint/data/ExtendedSpecs/) into [`CloudSpecs`](https://github.com/aws-cloudformation/cfn-python-lint/tree/master/src/cfnlint/data/CloudSpecs) and the ability to provide patterns in addition to literal strings 👍

Dislike JSON, separate `ResourceTypes`/`PropertyTypes` sections (should be inferable from type name alone and has caused confusion in the past: https://github.com/aws-cloudformation/cfn-python-lint/pull/1608/, https://github.com/aws-cloudformation/cfn-python-lint/pull/1654, https://github.com/aws-cloudformation/cfn-python-lint/pull/1712), and mapping [`Ref`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-ref.html)/[`GetAtt`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-getatt.html) per constraint

We should optimize for [concise constraints](https://github.com/aws-cloudformation/cfn-python-lint/issues/903#issuecomment-595356853) at the expense of potentially more one-time parsing code

```bash
# just an example possibility for discussion
StartingPosition if AWS::Lambda::EventSourceMapping.EventSourceArn == arn:aws*:(kinesis|kafka|dynamodb):*
```

File: src/cfnlint/data/AdditionalSpecs/RequiredBasedOnValue.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
+{
+    "PropertyTypes": {},
+    "ResourceTypes": {
+        "AWS::Lambda::EventSourceMapping": {
+            "EventSourceArn": [
+                {
+                    "Regex": "arn:aws:(kinesis|kafka|dynamodb):*",
```
Comment (PatMyron) [inline src/cfnlint/data/AdditionalSpecs/RequiredBasedOnValue.json:7]: [missing support for other partitions](https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html#arns-syntax)
```suggestion
                    "Regex": "arn:aws*:(kinesis|kafka|dynamodb):*",
```

**Comment by PatMyron** [comment]
Think investigating similar rules/requests will be helpful:
[CloudWatch](https://github.com/aws-cloudformation/cfn-python-lint/issues/1059), [DynamoDB](https://github.com/aws-cloudformation/cfn-python-lint/blob/master/src/cfnlint/rules/resources/dynamodb/BillingMode.py), [EC2](https://github.com/aws-cloudformation/cfn-python-lint/blob/master/src/cfnlint/rules/resources/ectwo/Ebs.py), [Lambda](https://github.com/aws-cloudformation/cfn-python-lint/issues/1439), [RDS](https://github.com/aws-cloudformation/cfn-python-lint/blob/master/src/cfnlint/rules/resources/rds/AuroraDBInstanceProperties.py), [RDS](https://github.com/aws-cloudformation/cfn-python-lint/blob/master/src/cfnlint/rules/resources/rds/AuroraScalingConfiguration.py), [Route53](https://github.com/aws-cloudformation/cfn-python-lint/blob/master/src/cfnlint/rules/resources/route53/HealthCheck.py), [Route53](https://github.com/aws-cloudformation/cfn-python-lint/blob/master/src/cfnlint/rules/resources/route53/RecordSet.py), [Route53](https://github.com/aws-cloudformation/cfn-python-lint/issues/94), [StepFunctions](https://github.com/aws-cloudformation/cfn-python-lint/blob/master/src/cfnlint/rules/resources/stepfunctions/StateMachine.py)

Something to note is that quite a few of them are the opposite: a certain property _cannot_ be specified based on the value of another property

File: src/cfnlint/data/AdditionalSpecs/RequiredBasedOnValue.json (hunk 1 above)
Comment (kddejong) [reply src/cfnlint/data/AdditionalSpecs/RequiredBasedOnValue.json]: I like the readability of what you provided but yea we will need to think about that and how to parse it.  Let me try a few things to see if I can make that work well.

Comment (PatMyron) [reply src/cfnlint/data/AdditionalSpecs/RequiredBasedOnValue.json]: Or maybe even just [schematize](https://json-schema.org/) some of these constraint files so we automatically catch mistakes

Comment (kddejong) [reply src/cfnlint/data/AdditionalSpecs/RequiredBasedOnValue.json]: I agree.  I removed the resource types and property types.  Let me schematize it and add testing for that file.

**Comment by kddejong** [comment]
For tracking after this PR is merged I will evaluate if these rules can be deleted and moved to this rule.
https://github.com/aws-cloudformation/cfn-python-lint/blob/master/src/cfnlint/rules/resources/rds/AuroraScalingConfiguration.py
https://github.com/aws-cloudformation/cfn-python-lint/blob/master/src/cfnlint/rules/resources/rds/AuroraDBInstanceProperties.py
https://github.com/aws-cloudformation/cfn-python-lint/blob/master/src/cfnlint/rules/resources/route53/HealthCheck.py

File: src/cfnlint/data/AdditionalSpecs/RequiredBasedOnValue.json (hunk 1 above)
Comment (kddejong) [reply src/cfnlint/data/AdditionalSpecs/RequiredBasedOnValue.json]: Created a JSON Schema for the file and added testing.

**Comment by kddejong** [comment]
@PatMyron I'm going to look at another rule for doing the opposite check because I'm finding use cases where it could be valuable.

File: test/unit/specs/test_required_based_on_value.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+LOGGER = logging.getLogger('cfnlint.maintenance')
+LOGGER.addHandler(logging.NullHandler())
+
+
+class TestRequiredBasedOnValueSpec(BaseTestCase):
+    """Used for Testing Rules"""
+
+
+    def test_success_sbd_domain_removed(self):
+        """Success removal of SBD Domain form unsupported regions"""
```
Comment (PatMyron) [inline test/unit/specs/test_required_based_on_value.py:21]: Some outdated copy-pasted code here
```suggestion


    def validate_additional_specs_schema(self):
```

File: test/unit/rules/resources/properties/test_required_based_on_value.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
+"""
+Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+SPDX-License-Identifier: MIT-0
+"""
+from test.unit.rules import BaseRuleTestCase
+from cfnlint.rules.resources.properties.RequiredBasedOnValue import RequiredBasedOnValue  # pylint: disable=E0401
+
+
+class TestRequiredBasedOnValue(BaseRuleTestCase):
+    """Test OnlyOne Property Configuration"""
```
Comment (PatMyron) [inline test/unit/rules/resources/properties/test_required_based_on_value.py:10]:
```suggestion
```

File: src/cfnlint/data/AdditionalSpecs/RequiredBasedOnValue.json (hunk 2 above)
Comment (PatMyron) [reply src/cfnlint/data/AdditionalSpecs/RequiredBasedOnValue.json]: @kddejong sorry, got confused by that trailing wildcard. Should be something like this instead:

```diff
+                    "Regex": "arn:aws.*:(kinesis|kafka|dynamodb):",
```

Comment (kddejong) [reply src/cfnlint/data/AdditionalSpecs/RequiredBasedOnValue.json]: Oops.  Resolved it here #1759

**Pull request merged**